pkg/syncer/upsync/upsync_controller.go

```diff
@@ -0,0 +1,437 @@
+	queue           workqueue.RateLimitingInterface
+	dirtyStatusKeys sync.Map
+
+	upstreamClient   kcpdynamic.ClusterInterface
+	downstreamClient dynamic.Interface
+
+	getDownstreamLister               func(gvr schema.GroupVersionResource) (cache.GenericLister, error)
+	listDownstreamNamespacesByLocator func(jsonLocator string) ([]*unstructured.Unstructured, error)
+	getUpstreamUpsyncerLister         func(gvr schema.GroupVersionResource) (kcpcache.GenericClusterLister, error)
+	getUpsyncedGVRs                   func() ([]schema.GroupVersionResource, error)
+
+	syncTargetName        string
+	syncTargetClusterName logicalcluster.Name
+	syncTargetUID         types.UID
+	syncTargetKey         string
+}
+
+// Queue handles keys for both upstream and downstream resources.
+type queueKey struct {
+	gvr schema.GroupVersionResource
+	key string
+}
+
+func (c *controller) enqueueUpstream(gvr schema.GroupVersionResource, obj interface{}, logger logr.Logger, dirtyStatus bool) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	logging.WithQueueKey(logger, key).V(2).Info("queueing", "gvr", gvr.String())
+	queueKey := queueKey{
+		gvr: gvr,
+		key: key,
+	}
+
+	if dirtyStatus {
+		c.dirtyStatusKeys.Store(queueKey, true)
+	}
+	c.queue.Add(queueKey)
+}
+
+func (c *controller) enqueueDeletedDownstreamNamespace(deletedNamespace *unstructured.Unstructured, logger logr.Logger) {
+	upstreamLocator, locatorExists, err := shared.LocatorFromAnnotations(deletedNamespace.GetAnnotations())
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	if !locatorExists || upstreamLocator == nil {
+		logger.V(4).Info("the namespace locator doesn't exist on deleted downstream namespace.")
+		return
+	}
+
+	if upstreamLocator.SyncTarget.UID != c.syncTargetUID || upstreamLocator.SyncTarget.ClusterName != c.syncTargetClusterName.String() {
+		return
+	}
+
+	upsyncedGVRs, err := c.getUpsyncedGVRs()
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	for _, gvr := range upsyncedGVRs {
+		upstreamLister, err := c.getUpstreamUpsyncerLister(gvr)
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+
+		upstreamUpsyncedResources, err := upstreamLister.ByCluster(upstreamLocator.ClusterName).ByNamespace(upstreamLocator.Namespace).List(labels.Everything())
+		if err != nil {
+			utilruntime.HandleError(err)
+			return
+		}
+
+		for _, upstreamUpsyncedResource := range upstreamUpsyncedResources {
+			c.enqueueUpstream(gvr, upstreamUpsyncedResource, logger, false)
+		}
+	}
+}
+
+func (c *controller) enqueueDownstream(gvr schema.GroupVersionResource, downstreamObj *unstructured.Unstructured, logger logr.Logger, dirtyStatus bool) {
+	downstreamNamespace := downstreamObj.GetNamespace()
+	locatorHolder := downstreamObj
+	if downstreamNamespace != "" {
+		downstreamNamespaceLister, err := c.getDownstreamLister(corev1.SchemeGroupVersion.WithResource("namespaces"))
```

Comment:
```suggestion
		// get locator from namespace for namespaced objects
		downstreamNamespaceLister, err := c.getDownstreamLister(corev1.SchemeGroupVersion.WithResource("namespaces"))
```

Reply:
Fixed in commit fffff6927c99c86702c7709881a77057331950cd